h2/src/test/org/h2/test/unit/TestSubqueryPerformanceOnLazyExecutionMode.java

```diff
@@ -0,0 +1,100 @@
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+/**
+ * Test subquery performance with lazy query execution mode {@link SetTypes#LAZY_QUERY_EXECUTION}.
+ */
+public class TestSubqueryPerformanceOnLazyExecutionMode extends TestBase {
+    /** Rows count. */
+    private static final int ROWS = 5000;
+
+    /**
+     * Run just this test.
+     *
+     * @param a ignored
+     */
+    public static void main(String[] a) throws Exception {
+        TestBase.createCaller().init().test();
+    }
+
+
+    @Override
+    public void test() throws Exception {
+        Class.forName("org.h2.Driver");
+        try (Connection conn = DriverManager.getConnection("jdbc:h2:mem:")) {
+            try (Statement stmt = conn.createStatement()) {
+                stmt.execute("CREATE TABLE ONE(X INTEGER , Y INTEGER )");
+
+                try (PreparedStatement prep = conn.prepareStatement("insert into one values(?,?)")) {
+                    for (int row = 0; row < ROWS; row++) {
+                        prep.setInt(1, row / 100);
+                        prep.setInt(2, row);
+                        prep.execute();
+                    }
+                }
+
+                testSubqueryInCondition(stmt);
+                testSubqueryInJoin(stmt);
+            }
+        }
+    }
+
+    public void testSubqueryInCondition(Statement stmt) throws Exception {
+        String sql = "SELECT COUNT (*) FROM one WHERE x IN (SELECT y FROM one WHERE y < 50)";
+
+        long tNotLazy = executeAndCheckResult(stmt, sql, false);
+        long tLazy = executeAndCheckResult(stmt, sql, true);
+
+        Assert.assertTrue("Lazy execution too slow. lazy time: "
+                        + tLazy + ", not lazy time: " + tNotLazy,
+                tNotLazy * 2 > tLazy);
+    }
+
+    public void testSubqueryInJoin(Statement stmt) throws Exception {
+        String sql =
+                "SELECT COUNT (one.x) FROM one " +
+                "JOIN (SELECT y AS val FROM one WHERE y < 50) AS subq ON subq.val=one.x";
+
+        long tNotLazy = executeAndCheckResult(stmt, sql, false);
+        long tLazy = executeAndCheckResult(stmt, sql, true);
+
+        Assert.assertTrue("Lazy execution too slow. lazy time: "
+                        + tLazy + ", not lazy time: " + tNotLazy,
+                tNotLazy * 2 > tLazy);
+    }
+
+    /**
+     * @return Time of the query execution.
+     */
+    private long executeAndCheckResult(Statement stmt, String sql, boolean lazy) throws SQLException {
+        if (lazy) {
+            stmt.execute("SET LAZY_QUERY_EXECUTION 1");
+        }
+        else {
+            stmt.execute("SET LAZY_QUERY_EXECUTION 0");
+        }
+
+        long t0 = System.currentTimeMillis();
+        try (ResultSet rs = stmt.executeQuery(sql)) {
+            rs.next();
+            Assert.assertEquals(ROWS, rs.getInt(1));
+        }
+
+        return System.currentTimeMillis() - t0;
```

Comment:
Unfortunately, on Travis VMs execution time is not reliable, unrelated delays are possible at any time. Test case must tolerate small amount of possible failures somehow.

Reply:
Sure, but the times of a lazy and not lazy modes must be almost the same.
Before the patch the execution times differ more than 10 times.
Times condition: NOT_LAZY_TIME * 5 > LAZY_TIME
I guess this is enough to make the test robust.

Your suggestions?

Reply:
May be in case of failure run the same statements multiple times and make sure that in most executions results were expected?

Reply:
Great, what about N repeats and the first appropriate execution time completes the test?
Only N failures of the time condition failure the test. Is it OK?

Reply:
If there are multiple failures it's better to fail the test, at least for now.

I think that if first execution was successful test should be passed. If not, a couple of additional executions can be performed and if they both were successful test should be passed too and failed otherwise.

At least it protects us against one random delay during the test.

We already have some tests that were disabled in Travis builds some time ago because we didn't find any way to make them stable.

Reply:
Thanks. Fixed.